### PR TITLE
TEST/MPI: Added MPI+CUDA example.

### PR DESCRIPTION
**Pull request by rakhmets**
## What?
Added MPI+CUDA example.

### CODE REVIEW

**Review by Akshay-Venkatesh — COMMENTED** [review]
@rakhmets Overall the tests stress multi-GPU support in a good way. 

I see the following not being tested:
1. The case where one thread has device context bound to it and it allocates device memory. Later another thread issues the MPI operations with the allocated memory. 
2. Also use cudaSetDevice/cudaDeviceReset runtime API Instead of explicitly using ctxRetain/Release. This is more commonly the API exercised by high-level applications so it would be good to ensure that no cases break from just testing driver API.

The above two cases are supported by multi-GPU support, right? If so, will they be tested in a separate tests?

File: test/mpi/test_mpi_cuda.c [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+    CUDA_CALL(cuDeviceGetCount(&dev_count));
+    if (dev_count < 1) {
+        PRINT_ROOT("This test requires at least 1 GPU\n");
+        goto out;
+    }
+
+    for (dev_idx = dev_count - 1; dev_idx > -1; --dev_idx) {
+        CUDA_CALL(cuDeviceGet(&cu_dev, dev_idx));
+        CUDA_CALL(cuDevicePrimaryCtxRetain(&cu_ctx, cu_dev));
+    }
```
Comment (Akshay-Venkatesh) [inline test/mpi/test_mpi_cuda.c:469]: @rakhmets Minor comment -- By this logic, if the MPI job were to be launched on the same node, both ranks would end up initializing cu_dev to 0 by the end of the above loop. If there are multiple GPUs on the node, I think it would be good to check where the ranks manage different GPUs. As an example, osu benchmarks use LOCAL_RANK env for this reason.

**Comment by rakhmets** [comment]
> @rakhmets Overall the tests stress multi-GPU support in a good way.
> 
> I see the following not being tested:
> 
> 1. The case where one thread has device context bound to it and it allocates device memory. Later another thread issues the MPI operations with the allocated memory.
> 2. Also use cudaSetDevice/cudaDeviceReset runtime API Instead of explicitly using ctxRetain/Release. This is more commonly the API exercised by high-level applications so it would be good to ensure that no cases break from just testing driver API.
> 
> The above two cases are supported by multi-GPU support, right? If so, will they be tested in a separate tests?

1. `test_alloc_prim_send_no` does pretty much the same thing (from the perspective of the feature implementation). There is an active (retained) primary device context, but it is not bound to the thread at the moment of MPI send/recv.
2. I think it would be better to have separate test for CUDA Runtime API only. Because some scenarios (e.g. create user context) are only valid for Driver API.

I will add a separate test using CUDA Runtime API. And I will include the test case described in the first bullet to this new test.

File: contrib/test_jenkins.sh [hunk 2]
```diff
@@ -672,16 +672,18 @@ run_mpi_tests() {
...
-			build release --disable-gtest --with-mpi
+			build release-mt --with-mpi
 
 			# check whether installation is valid (it compiles examples at least)
 			$MAKEP installcheck
 
-			MPIRUN="mpirun \
+			MPIRUN_COMMON="mpirun \
 					--allow-run-as-root \
 					--bind-to none \
```
Comment (michal-shalev) [inline contrib/test_jenkins.sh:682]: maybe align?

File: test/mpi/test_mpi_cuda.c [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+#define PRINT_ROOT(fmt, ...) \
+    do { \
+        int _rank; \
+        MPI_Comm_rank(MPI_COMM_WORLD, &_rank); \
+        if (_rank == 0) { \
+            fprintf(stdout, fmt, ##__VA_ARGS__); \
+        } \
+    } while (0)
+
+#define array_size(_array) (sizeof(_array) / sizeof((_array)[0]))
```
Comment (michal-shalev) [inline test/mpi/test_mpi_cuda.c:35]: if you use UCS anyways, can you use the ucs macro instead?

File: test/mpi/test_mpi_cuda.c [hunk 4]
```diff
@@ -0,0 +1,668 @@
...
+    thread_arg.d_send = alloc_mem_send.ptr;
+    thread_arg.d_recv = alloc_mem_recv.ptr;
+    thread_arg.size   = size;
+    pthread_create(&thread, NULL, mpi_pingpong_thread, &thread_arg);
+    pthread_join(thread, NULL);
+
+    res = check_result(alloc_mem_recv.ptr, gold_data, size);
+
+    allocator_send->free(&alloc_mem_send);
+    allocator_recv->free(&alloc_mem_recv);
```
Comment (michal-shalev) [inline test/mpi/test_mpi_cuda.c:406]: maybe a function to free?

File: test/mpi/test_mpi_cuda.c [hunk 5]
```diff
@@ -0,0 +1,668 @@
...
+
+    return res;
+}
+
+static int test_alloc_prim_send_other_prim(const allocator_t *allocator_send,
+                                           const allocator_t *allocator_recv,
+                                           const void *gold_data, size_t size,
+                                           CUdevice cu_dev)
+{
+    int res = 0, rank, dev_count, i;
```
Comment (michal-shalev) [inline test/mpi/test_mpi_cuda.c:482]: code style

File: test/mpi/test_mpi_cuda.c [hunk 6]
```diff
@@ -0,0 +1,668 @@
...
+
+        pop_and_release_primary_context(cu_dev_comm);
+
+        res += check_result(alloc_mem_recv.ptr, gold_data, size);
+
+        allocator_send->free(&alloc_mem_send);
+        allocator_recv->free(&alloc_mem_recv);
+
+        pop_and_release_primary_context(cu_dev_alloc);
+    }
```
Comment (michal-shalev) [inline test/mpi/test_mpi_cuda.c:511]: maybe add the GPU index in a log?

File: test/mpi/test_mpi_cuda.c [hunk 7]
```diff
@@ -0,0 +1,668 @@
...
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <mpi.h>
+#include <cuda.h>
+#include <pthread.h>
+
+#define SIZE_S 8
+#define SIZE_M 4 * 1024 * 1024
+#define SIZE_L 16 * 1024 * 1024
```
Comment (michal-shalev) [inline test/mpi/test_mpi_cuda.c:21]: Missing parentheses?

File: test/mpi/test_mpi_cuda.c [hunk 8]
```diff
@@ -0,0 +1,668 @@
...
+    {"MPI pingpong, memory is allocated with the primary context set, "
+     "communication is done with the user context set",
+     1, 0, test_alloc_prim_send_user},
+    {"MPI pingpong, memory is allocated with the user context set, "
+     "communication is done with the primary context set",
+     1, 0, test_alloc_user_send_prim},
+    {"MPI pingpong, memory is allocated with the primary context of one GPU, "
+     "communication is done with the primary contexts of another GPU",
+     2, 0, test_alloc_prim_send_other_prim}
+};
```
Comment (michal-shalev) [inline test/mpi/test_mpi_cuda.c:535]: IMO this should appear after the `test_t` struct definition

File: test/mpi/test_mpi_cuda.c [hunk 9]
```diff
@@ -0,0 +1,668 @@
...
+    cuda_memcpy((void*)alloc_mem_send.ptr, gold_data, size);
+
+    mpi_pingpong(alloc_mem_send.ptr, alloc_mem_recv.ptr, size);
+
+    res = check_result(alloc_mem_recv.ptr, gold_data, size);
+
+    allocator_send->free(&alloc_mem_send);
+    allocator_recv->free(&alloc_mem_recv);
+
+    pop_and_release_primary_context(cu_dev);
```
Comment (michal-shalev) [inline test/mpi/test_mpi_cuda.c:343]: This part repeats 4 times in:
1. test_alloc_prim_send_prim
2. test_alloc_prim_send_no
3. test_alloc_prim_send_thread
4. test_alloc_prim_send_user

IMO you can create a static helper function for check result + free + pop primary context

Comment (michal-shalev) [reply test/mpi/test_mpi_cuda.c:343]: And I think it should call another static func for free, which repeats in all 6 tests functions

File: test/mpi/test_mpi_cuda.c [hunk 10]
```diff
@@ -0,0 +1,668 @@
...
+        int _rank; \
+        MPI_Comm_rank(MPI_COMM_WORLD, &_rank); \
+        if (_rank == 0) { \
+            fprintf(stdout, fmt, ##__VA_ARGS__); \
+        } \
+    } while (0)
+
+#define array_size(_array) (sizeof(_array) / sizeof((_array)[0]))
+
+#define CUDA_CALL(_func) \
```
Comment (michal-shalev) [inline test/mpi/test_mpi_cuda.c:37]: Maybe rename to `CUDA_CALL_ASSERT`? `CUDA_CALL` does not clearly imply that it can exit the program

File: test/mpi/test_mpi_cuda.c [hunk 11]
```diff
@@ -0,0 +1,668 @@
...
+                        MPI_Reduce(MPI_IN_PLACE, &res, 1, MPI_INT, MPI_SUM, 0,
+                                   MPI_COMM_WORLD);
+                    } else {
+                        MPI_Reduce(&res, NULL, 1, MPI_INT, MPI_SUM, 0,
+                                   MPI_COMM_WORLD);
+                    }
+
+                    PRINT_ROOT(TEST_NAME_SIZE_FMT ": %s\n", i,
+                               allocator_send->name, allocator_recv->name,
+                               sizes[l], (res ? "FAIL" : "PASS"));
```
Comment (michal-shalev) [inline test/mpi/test_mpi_cuda.c:649]: IMO a helper function like `run_test_for_all_sizes()` would help here becasue main() is long and has deep nesting

File: test/mpi/test_mpi_cuda.c [hunk 12]
```diff
@@ -0,0 +1,668 @@
...
+    CUcontext cu_ctx;
+    void *gold_data;
+    int i, j, k, l;
+    const test_t *test;
+    const allocator_t *allocator_send, *allocator_recv;
+    int res;
+
+    MPI_Init_thread(&argc, &argv, required, &provided);
+    MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
+    if (comm_size != 2) {
```
Comment (michal-shalev) [inline test/mpi/test_mpi_cuda.c:574]: IMO the number of processes should be in a define

File: test/mpi/test_mpi_cuda.c [hunk 13]
```diff
@@ -0,0 +1,668 @@
...
+        PRINT_ROOT("This test requires 2 processes, not %d\n", comm_size);
+        goto out;
+    }
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    CUDA_CALL(cuInit(0));
+    CUDA_CALL(cuDeviceGetCount(&dev_count));
+    if (dev_count < 1) {
```
Comment (michal-shalev) [inline test/mpi/test_mpi_cuda.c:583]: IMO the minimum GPUs for this test should be a define, and then you can use it to set the field in `test_t`

File: test/mpi/test_mpi_cuda.c [hunk 14]
```diff
@@ -0,0 +1,668 @@
...
+    int required = MPI_THREAD_SINGLE;
+#endif
+    int provided;
+    int comm_size;
+    int rank;
+    int dev_count, dev_idx;
+    CUdevice cu_dev;
+    CUcontext cu_ctx;
+    void *gold_data;
+    int i, j, k, l;
```
Comment (michal-shalev) [inline test/mpi/test_mpi_cuda.c:567]: IMO it would be more readable to use:
I -> test_idx
j -> alloc_send_idx
k -> alloc_recv_idx
l -> size_idx
becasue main() is long and has deep nesting

File: contrib/test_jenkins.sh [hunk 15]
```diff
@@ -672,16 +672,18 @@ run_mpi_tests() {
 			save_LD_LIBRARY_PATH=${LD_LIBRARY_PATH}
 			export LD_LIBRARY_PATH=${ucx_inst}/lib:${MPI_HOME}/lib:${prev_LD_LIBRARY_PATH}
 
-			build release --disable-gtest --with-mpi
+			build release-mt --with-mpi
```
Comment (brminich) [inline contrib/test_jenkins.sh:675]: i'd include asserts

File: contrib/test_jenkins.sh [hunk 16]
```diff
@@ -693,6 +695,12 @@ run_mpi_tests() {
 
 			test_malloc_hooks_mpi
 
+			if [ "X$have_cuda" == "Xyes" ] && [ -x ./test/mpi/test_mpi_cuda ]
+			then
+				echo "==== Running MPI CUDA tests ===="
+				${MPIRUN_COMMON} -np 2 ./test/mpi/test_mpi_cuda
```
Comment (brminich) [inline contrib/test_jenkins.sh:701]: can we run it several times with different env vars? E.g.
- UCX_TLS=^cuda_ipc // to test cuda_copy
- UCX_RNDV_SHEME=put_ppln
maybe some others?

File: test/mpi/test_mpi_cuda.c [hunk 17]
```diff
@@ -0,0 +1,668 @@
...
+
+#define array_size(_array) (sizeof(_array) / sizeof((_array)[0]))
+
+#define CUDA_CALL(_func) \
+    do { \
+        CUresult __err = (_func); \
+        if (__err != CUDA_SUCCESS) { \
+            const char *__err_str; \
+            cuGetErrorString(__err, &__err_str); \
+            fprintf(stderr, "test_mpi_cuda.c:%-3u %s failed: %d (%s)\n", \
```
Comment (brminich) [inline test/mpi/test_mpi_cuda.c:43]: can you use `__FILE__` instead?

File: test/mpi/test_mpi_cuda.c [hunk 18]
```diff
@@ -0,0 +1,668 @@
...
+    const char     *description;
+    const unsigned min_gpus;
+    const int      mthread_support;
+    int (*func)(const allocator_t*, const allocator_t*, const void*, size_t,
+                CUdevice);
+} test_t;
+
+typedef struct {
+    CUdeviceptr d_send;
+    CUdeviceptr d_recv;
```
Comment (brminich) [inline test/mpi/test_mpi_cuda.c:71]: minor: maybe use `send_ptr` and `recv_ptr` or use some other consistent names for `CUdeviceptr` variables

File: test/mpi/test_mpi_cuda.c [hunk 19]
```diff
@@ -0,0 +1,668 @@
...
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    receiver = (sender + 1) % 2;
+
+    if (rank == sender) {
+        MPI_Send((void*)d_send, size, MPI_BYTE, receiver, 0, MPI_COMM_WORLD);
+    } else {
+        MPI_Recv((void*)d_recv, size, MPI_BYTE, sender, 0, MPI_COMM_WORLD,
+                 MPI_STATUS_IGNORE);
```
Comment (brminich) [inline test/mpi/test_mpi_cuda.c:123]: maybe it is worth checking return values here?

File: test/mpi/test_mpi_cuda.c [hunk 20]
```diff
@@ -0,0 +1,668 @@
...
+    CUDA_CALL(cuMemRelease((CUmemGenericAllocationHandle)alloc_mem->obj));
+}
+
+#if HAVE_CUDA_FABRIC
+static alloc_mem_t alloc_vmm_fabric(size_t size)
+{
+    return alloc_vmm_type(size, CU_MEM_HANDLE_TYPE_FABRIC);
+}
+
+static alloc_mem_t alloc_mempool(size_t size)
```
Comment (brminich) [inline test/mpi/test_mpi_cuda.c:251]: can we have mempool without FABRIC?
Or do we have FABRIC cap on our CI machines?

File: test/mpi/test_mpi_cuda.c [hunk 21]
```diff
@@ -0,0 +1,668 @@
...
+{
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    int res;
+
+    retain_and_push_primary_context(cu_dev);
+
+    alloc_mem_send = allocator_send->alloc(size);
+    alloc_mem_recv = allocator_recv->alloc(size);
+
+    cuda_memcpy((void*)alloc_mem_send.ptr, gold_data, size);
```
[the remark this answers is not in the document]
Comment (brminich) [reply test/mpi/test_mpi_cuda.c:334]: in general seems it is possible to avoid some code duplications

Comment (brminich) [inline test/mpi/test_mpi_cuda.c:334]: maybe it is worth to memset recv buffer to zeros or something

File: test/mpi/test_mpi_cuda.c [hunk 22]
```diff
@@ -0,0 +1,668 @@
...
+
+    CUDA_CALL(cuCtxPopCurrent(&primary_ctx));
```
Comment (brminich) [inline test/mpi/test_mpi_cuda.c:364]: can we add a check that there is no other context set currently?

File: test/mpi/test_mpi_cuda.c [hunk 23]
```diff
@@ -0,0 +1,668 @@
...
+{
+    int res = 0, rank, dev_count, i;
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    CUdevice cu_dev_alloc, cu_dev_comm;
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    CUDA_CALL(cuDeviceGetCount(&dev_count));
+
+    for (i = 0; i < dev_count; ++i) {
+        CUDA_CALL(cuDeviceGet(&cu_dev_alloc, (i + rank) % dev_count));
```
Comment (brminich) [inline test/mpi/test_mpi_cuda.c:490]: maybe iterate over all possible combinations instead?

File: test/mpi/test_mpi_cuda.c [hunk 24]
```diff
@@ -0,0 +1,668 @@
...
+                allocator_recv = allocators + k;
+                if (!check_allocator(allocator_recv, cu_dev)) {
+                    PRINT_ROOT(TEST_NAME_FMT ": SKIP (not supported)\n", i,
+                               allocator_send->name, allocator_recv->name);
+                    continue;
+                }
+
+                for (l = 0; l < array_size(sizes); ++l) {
+                    total_ucx_errors_and_warnings = 0;
+                    res = test->func(allocator_send, allocator_recv, gold_data,
```
Comment (brminich) [inline test/mpi/test_mpi_cuda.c:635]: why are we using gold_data of size L for all sizes?

File: contrib/test_jenkins.sh (hunk 2 above)
Comment (rakhmets) [reply contrib/test_jenkins.sh:682]: I don't see a single alignment rule in this file. I left the previous alignment in two tabs.

File: test/mpi/test_mpi_cuda.c (hunk 17 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c:43]: It will print full path which is not needed here. And I don't want to add code to fetch a filename from the __FILE__, since it overcomplicates simple code.

File: test/mpi/test_mpi_cuda.c (hunk 24 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c:635]: Because gold_data of size S is equal to first S bytes of gold_data of size M, and gold_data of size M is equal to first M bytes of gold_data of size L.

Comment (brminich) [reply test/mpi/test_mpi_cuda.c:635]: right, i somehow thought there is larger value

File: contrib/test_jenkins.sh (hunk 2 above)
Comment (rakhmets) [reply contrib/test_jenkins.sh:682]: I don't see a single alignment rule in this file. I left the previous alignment in two tabs.

File: test/mpi/test_mpi_cuda.c (hunk 3 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: Fixed.

File: test/mpi/test_mpi_cuda.c [hunk 25]
```diff
@@ -0,0 +1,668 @@
...
+    int i;
+
+    gold_data = malloc(size);
+    if (gold_data == NULL) {
+        fprintf(stderr, "failed to create gold data\n");
+        exit(-1);
+    }
+
+    for (i = 0; i < size; ++i) {
+        gold_data[i] = i % 256;
```
[the remark this answers is not in the document]
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: Done.

File: test/mpi/test_mpi_cuda.c (hunk 5 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: Fixed.

File: test/mpi/test_mpi_cuda.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: Fixed.

File: test/mpi/test_mpi_cuda.c (hunk 10 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: Renamed to `CUDA_CHECK`.

File: test/mpi/test_mpi_cuda.c (hunk 13 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: Defined a value.

File: test/mpi/test_mpi_cuda.c (hunk 17 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c:44]: It will print full path which is not needed here. And I don't want to add code to fetch a filename from the FILE, since it overcomplicates simple code.

File: test/mpi/test_mpi_cuda.c (hunk 18 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: Renamed the fields.

File: test/mpi/test_mpi_cuda.c (hunk 20 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c:308]: I don't think that we need mempool without fabric feature in this test, since it won't check any code path, that isn't covered by existing allocators.

File: test/mpi/test_mpi_cuda.c (hunk 21 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: Done.

File: test/mpi/test_mpi_cuda.c (hunk 22 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: Added.

File: test/mpi/test_mpi_cuda.c (hunk 19 above)
Comment (brminich) [reply test/mpi/test_mpi_cuda.c:178]: to make sure we do not miss an error if it silently fails

File: test/mpi/test_mpi_cuda.c (hunk 9 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: Added function, which checks result and frees memory. Keep releasing primary context outside the method, because test_alloc_prim_send_user has the different sequence of calls, and it more obvious to see what each test does.

File: test/mpi/test_mpi_cuda.c (hunk 23 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: Left only one compbination, since the test allocates on one gpu and send using another. Iterating through all gpus doesn't improve coverage.

File: contrib/test_jenkins.sh (hunk 16 above)
Comment (rakhmets) [reply contrib/test_jenkins.sh:701]: Added test variants.

File: test/mpi/test_mpi_cuda.c (hunk 17 above)
Comment (brminich) [reply test/mpi/test_mpi_cuda.c:44]: will it be just something like below?
```
const char *filename = strrchr(__FILE__, '/');
filename = filename ? filename + 1 : __FILE__;
```
imo, it is better, so no need to change the code if file is renamed for some reason. Another alternative is just do not print a file name, because it is anyway a single file test

File: test/mpi/test_mpi_cuda.c (hunk 20 above)
Comment (brminich) [reply test/mpi/test_mpi_cuda.c:308]: is FABRIC currently checked on our CI machines?

File: test/mpi/test_mpi_cuda.c [hunk 26]
```diff
@@ -0,0 +1,695 @@
...
+    alloc_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    CUDA_CHECK(cuCtxPopCurrent(&primary_ctx));
+    CUDA_CHECK(cuCtxCreate(&user_ctx, 0, params->cu_dev));
+
+    mpi_pingpong(params->rank, alloc_mem_send.ptr, alloc_mem_recv.ptr,
+                 params->size);
+
+    CUDA_CHECK(cuCtxDestroy(user_ctx));
+    CUDA_CHECK(cuCtxPushCurrent(primary_ctx));
```
Comment (brminich) [inline test/mpi/test_mpi_cuda.c:486]: can we also free with user ctx?

File: test/mpi/test_mpi_cuda.c [hunk 27]
```diff
@@ -0,0 +1,695 @@
...
+    for (dev_idx = dev_count - 1; dev_idx > -1; --dev_idx) {
+        CUDA_CHECK(cuDeviceGet(&cu_dev, dev_idx));
+        CUDA_CHECK(cuDevicePrimaryCtxRetain(&cu_ctx, cu_dev));
+    }
+
+    if ((test_params.rank == 1) && (dev_count > 1)) {
+        CUDA_CHECK(cuDeviceGet(&cu_dev, 1));
+    }
+
+    test_params.gold_data = create_gold_data(SIZE_L);
```
Comment (brminich) [inline test/mpi/test_mpi_cuda.c:633]: can gold_data be just a global ptr, so there will be no need to pass it in params down till check func?

File: test/mpi/test_mpi_cuda.c [hunk 28]
```diff
@@ -0,0 +1,695 @@
...
+    pop_and_release_primary_context(params->cu_dev);
+
+    res = check_and_free_mem(params, &alloc_mem_send, &alloc_mem_recv);
+
+    CUDA_CHECK(cuCtxDestroy(user_ctx));
+
+    return res;
+}
+
+static int test_alloc_prim_send_other_prim(const test_params_t *params)
```
Comment (brminich) [inline test/mpi/test_mpi_cuda.c:517]: can we do send/recv between all possible combinations of available gpus?

File: contrib/test_jenkins.sh (hunk 15 above)
Comment (rakhmets) [reply contrib/test_jenkins.sh]: Added `--enable-assertions`

File: test/mpi/test_mpi_cuda.c (hunk 17 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: Removed.

File: test/mpi/test_mpi_cuda.c (hunk 26 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c:485]: Yes,  we can. But why are you asking? Would you like to change the test?
All the tests have the same pattern currently.
```
1. set context
2. allocate data
3. change context
4. transfer data
5. revert context change
6. free data
7. unset context
```

File: test/mpi/test_mpi_cuda.c (hunk 27 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c]: made it a global var.

File: test/mpi/test_mpi_cuda.c (hunk 28 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c:514]: Yes, we can. However, it doesn't bring any additional value for the test. The point is to test the following scenario:
```
1. set device X
2. allocate data
3. set device Y
4. send data
5. unset device Y
6. free data
7. unset device X
```

File: test/mpi/test_mpi_cuda.c (hunk 20 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c:305]: No, we don't have FABRIC capability on our CI machines, and it is not checked there. Mempool allocator acts like a mempool without fabric handle type.

Comment (brminich) [reply test/mpi/test_mpi_cuda.c:305]: but we define mempool with just FABRIC handles, no? How do we check it without fabric?

File: test/mpi/test_mpi_cuda.c (hunk 26 above)
Comment (brminich) [reply test/mpi/test_mpi_cuda.c:483]: yes, i'm afraid that we could miss some corner case, especially given that all the tests doing the same pattern

File: test/mpi/test_mpi_cuda.c (hunk 28 above)
Comment (brminich) [reply test/mpi/test_mpi_cuda.c:514]: considering that we may want to run this test also on non-CI machines (possibly DGX-like), running all-2-all  seems to be a better coverage. wdyt?

Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c:514]: I don't think that this will improve coverage, or I don't see the case. Could you please explain the scenario that will be covered by these changes.

File: test/mpi/test_mpi_cuda.c (hunk 26 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c:483]: Would you like to suggest a scenario that needs to be added?

File: test/mpi/test_mpi_cuda.c (hunk 20 above)
Comment (rakhmets) [reply test/mpi/test_mpi_cuda.c:305]: Yes. The handle type is ignored in our CI.

File: test/mpi/test_mpi_cuda.c (hunk 28 above)
Comment (brminich) [reply test/mpi/test_mpi_cuda.c:514]: some configurations, like DGX v1 may have different communication properties between different GPUs
you can add todo for later addition for now

File: test/mpi/test_mpi_cuda.c (hunk 26 above)
Comment (brminich) [reply test/mpi/test_mpi_cuda.c:483]: i was thinking of some, but seems it is mostly testing cuda rather than UCX, so no need

**Review by brminich — APPROVED** [review]

**Review by michal-shalev — APPROVED** [review]

**Comment by brminich** [comment]
@rakhmets, can u pls squash?
